lib/ansible/plugins/connection/buildah.py

```diff
@@ -0,0 +1,115 @@
+class Connection(ConnectionBase):
+    """
+    This is a connection plugin for buildah: it uses buildah binary to interact with the containers
+    """
+
+    # String used to identify this Connection class from other classes
+    transport = 'buildah'
+    has_pipelining = True
+    # TODO: add support to change user in the container
+    become_methods = frozenset(C.BECOME_METHODS)
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(play_context, new_stdin, *args, **kwargs)
+
+        self._container_id = self._play_context.remote_addr
+        self._connected = False
+        # if self._play_context.remote_user is not None:
+        # TODO: save actual user and print it
+
+    def _connect(self):
+        """ Create a container from specified container image, via host """
+        super(Connection, self)._connect()
+        self._connected = True
+
+    @ensure_connect
+    def exec_command(self, cmd, in_data=None, sudoable=False):
+        """ Run a command on the docker host """
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        display.v("%s" % self.connected)
+
+        cmd_bytes = to_bytes(cmd, errors='surrogate_or_strict')
+        cmd_args_list = shlex.split(cmd_bytes)
+
+        local_cmd = ['buildah', 'run', '--', self._container_id]
+        local_cmd += cmd_args_list
+        local_cmd = [to_bytes(i, errors='surrogate_or_strict') for i in local_cmd]
+
+        display.vvv("RUN %s" % (local_cmd,), host=self._container_id)
+        p = subprocess.Popen(local_cmd, shell=False, stdin=subprocess.PIPE,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        stdout, stderr = p.communicate(in_data)
+        stdout = to_bytes(stdout, errors='surrogate_or_strict')
+        stderr = to_bytes(stderr, errors='surrogate_or_strict')
+        display.vvvvv("STDOUT %s STDERR %s" % (stderr, stderr))
+        return (p.returncode, stdout, stderr)
+
+    def put_file(self, in_path, out_path):
+        """ Place a local file located in in_path inside container on out_path """
+        super(Connection, self).put_file(in_path, out_path)
+        display.vvv("PUT %s TO %s" % (in_path, out_path), host=self._container_id)
+
+        local_cmd = ['buildah', 'copy', '--', self._container_id, in_path, out_path]
+        local_cmd = [to_bytes(i, errors='surrogate_or_strict') for i in local_cmd]
+
+        display.vvv("RUN %s" % (local_cmd,), host=self._container_id)
+        p = subprocess.Popen(local_cmd, shell=False, stdin=subprocess.PIPE,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        stdout, stderr = p.communicate()
+        stdout = to_bytes(stdout, errors='surrogate_or_strict')
+        stderr = to_bytes(stderr, errors='surrogate_or_strict')
+        display.vvvvv("STDOUT %s STDERR %s" % (stdout, stderr))
+
+    def fetch_file(self, in_path, out_path):
```

Comment:
I don't understand how shells have anything to do with this: it's not even invoked via shell. The sample I posted above is a shortcut to:

```python
local_cmd = ['buildah', 'run', '--', self._container_id, 'cat',
             to_bytes(in_path, errors='surrogate_or_strict')]
with open(to_bytes(out_path, errors='surrogate_or_strict'), 'wb') as out_file:
    p = subprocess.Popen(local_cmd, shell=False, stdin=subprocess.PIPE,
                         stdout=out_file, stderr=subprocess.PIPE)
```

which I hoped would be obvious, sorry for not being precisely specific.

I understand that encodings might get messy. I'll take a look on the other workarounds. I fear that doing it with `dd` would not be possible since `dd` would get executed in the container and we need to get the file to host (some mounting magic could do it).